### PR TITLE
Refactor: Workspace factory model

### PR DESCRIPTION
**Pull request by chynu**
Refactored all functions from `wfactory_model.js` into refactor files.

For future: implement functions, clean up code (remove functions that are not necessary).

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly-devtools/80)

### CODE REVIEW

File: src/controller/toolbox_controller.js [hunk 1]
```diff
@@ -108,6 +108,23 @@ class ToolboxController {
...
+     *
+     * References:
+     * - transferFlyoutBlocksToCategory()
+     * - ListElement
+     * - addElementToList()
+     * - addSeparatorTab()
+     * - switchElement()
+     * - updatePreview()
+     */
+  }
```
Comment (AnmAtAnm) [inline src/controller/toolbox_controller.js]: Console.warn("unimplemented ToolboxController.addSeparator()")

File: src/controller/toolbox_controller.js [hunk 2]
```diff
@@ -108,6 +108,23 @@ class ToolboxController {
   }
 
   /**
+   * Adds separator to current Toolbox.
```
Comment (AnmAtAnm) [inline src/controller/toolbox_controller.js]: Looking at what this does in the production dev tools, let's rename this addCategorySeparator() (and update the doc).

FYI: You can also have separators between blocks within a toolbox category flyout.  Added issue #81 to track this issue (easy stretch goal).

File: src/model/toolbox.js [hunk 3]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * Return ordered list of ListElement objects in this instance of Toolbox.
+   * @return {!Array.<!ListElement>} ordered list of ListElement objects
    */
-  addBlock(categoryName, blockName, xmlDefinition) {
-    console.log('Add block called!');
-    // TODO: Implement function.
+  getCategoryList() {
+    /*
+     * TODO: Move in from wfactory_model.js:getToolboxList()
+     */
```
Comment (AnmAtAnm) [inline src/model/toolbox.js:105]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 4]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+     * TODO: Move in from wfactory_model.js:clearToolboxList()
+     *
+     * References:
+     * - this.toolboxList
+     * - this.hasVariablesCategory
+     * - this.hasProceduresCategory
+     * - this.shadowBlocks
+     * - this.selected
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented"

File: src/model/toolbox.js [hunk 5]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+    /*
+     * TODO: Move in from wfactory_model.js:addElementToList(element)
+     *
+     * References:
+     * - this.hasVariableCategory
+     * - this.hasProcedureCategory
+     * - this.toolboxList
+     * - this.flyout
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throws "unimplemented ..."

File: src/model/toolbox.js [hunk 6]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   */
+  deleteElement(index) {
+    /*
+     * TODO: Move in from wfactory_model.js:deleteElementFromList(index)
+     *
+     * References:
+     * - this.toolboxList
+     * - this.hasVariableCategory
+     * - this.hasProcedureCategory
+     */
```
Comment (AnmAtAnm) [inline src/model/toolbox.js:155]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 7]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   */
+  moveElement(element, newIndex, oldIndex) {
+    /*
+     * TODO: Move in from wfactory_model.js:moveElementToIndex(element, newIndex, oldIndex)
+     *
+     * References:
+     * - this.toolboxList
+     * - deleteElementFromList(oldIndex)
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 8]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   *     null if that element does not exist.
+   */
+  getElementById(id) {
+    /*
+     * TODO: Move in from wfactory_model.js
+     *
+     * References:
+     * - this.toolboxList
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 9]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * @return {ListElement} The corresponding ListElement object in toolboxList.
+   */
+  getElementByIndex(index) {
+    /*
+     * TODO: Move in from wfactory_model.js
+     *
+     * References:
+     * - this.toolboxList
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 10]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * element.
+   * @param {string} id ID of list element that should now be selected.
+   */
+  setSelected(id) {
+    /*
+     * TODO: Move in from wfactory_model.js:setSelectedById(id)
+     *
+     * References: N/A
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 11]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+  /**
+   * Returns ID of the currently selected element. Returns null if there are
+   * no categories (if selected == null).
+   * @return {string} The ID of the element currently selected.
+   */
+  getSelectedId() {
+    /*
+     * TODO: Move in from wfactory_model.js
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 12]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * are no category (if selected == null) or the selected element is not
+   * a category (in which case its name is null).
+   *
+   * @return {string} The name of the category currently selected.
+   */
+  getSelectedName() {
+    /*
+     * TODO: Move in from wfactory_model.js:getSelectedName()
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 13]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+  /**
+   * Returns the XML to load the selected element.
+   * @return {!Element} The XML of the selected element, or null if there is
+   *     no selected element.
+   */
+  getSelectedXml() {
+    /*
+     * TODO: Move in from wfactory_model.js
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 14]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   *     doesn't exist.
+   */
+  getElementIndex(id) {
+    /*
+     * TODO: Move in from wfactory_model.js:getIndexByElementId(id)
+     *
+     * References:
+     * - this.toolboxList
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 15]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * @return {number} ID of category
+   */
+  getCategoryId(name) {
+    /*
+     * TODO: Move in from wfactory_model.js:getCategoryIdByName(name)
+     *
+     * References:
+     * - this.toolboxList
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 16]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   */
+  addCustomTag(category, tag) {
+    /*
+     * TODO: Move in from wfactory_model.js
+     *
+     * References:
+     * - this.hasVariablesCategory
+     * - this.hasProcedureCategory
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 17]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+     *
+     * References:
+     * - this.toolboxList
+     * - this.flyout
+     * - ListElement
+     * - this.selected
+     */
+
+    // TODO: Rename function to be more clear/readable.
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 18]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   * @param {string} blockId The unique ID of block to be removed.
+   */
+  removeShadowBlock(blockId) {
+    /*
+     * TODO: Move in from wfactory_model.js
+     *
+     * References:
+     * - this.shadowBlocks
+     */
+  }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: console.warn or throw "unimplemented ..."

File: src/model/toolbox.js [hunk 19]
```diff
@@ -33,39 +37,330 @@ class Toolbox extends Resource {
...
+   *     otherwise.
+   */
+  isShadowBlock(blockId) {
+    /*
+     * TODO: Move in from wfactory_model.js
+     *
+     * References:
+     * - this.shadowBlocks
+     */
   }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/model/toolbox.js [hunk 20]
```diff
@@ -91,44 +386,64 @@ class Toolbox extends Resource {
...
     /*
-     * TODO: implement
+     * TODO: Move in from wfactory_model.js:hasCategoryByName
      *
-     * References: N/A
+     * References:
+     * - ListElement
      */
-    throw "unimplemented: setName";
   }
```
Comment (AnmAtAnm) [inline src/model/toolbox.js]: throw "unimplemented ..."

File: src/view/toolbox_editor_view.js [hunk 21]
```diff
@@ -221,22 +221,12 @@ class ToolboxEditorView {
      */
   }
 
-  /**
-   * If categories already exist, adds a separator to the model and view.
-   * Does not switch to select the separator, and updates the preview.
-   */
-  addSeparator() {
+  getShadowBlocksInWorkspace() {
```
Comment (AnmAtAnm) [inline src/view/toolbox_editor_view.js]: JSDoc

File: src/view/toolbox_editor_view.js [hunk 22]
```diff
@@ -221,22 +221,12 @@ class ToolboxEditorView {
...
-     * - addSeparatorTab()
-     * - addClickToSwitch()
-     * - switchElement()
-     * - updatePreview()
+     * TODO: Move in from wfactory_model.js
      */
+
+    // TODO: Investigate if this function is necessary at all. Possibly delete.
+    //       (This method was never called in DevTools)
   }
```
Comment (AnmAtAnm) [inline src/view/toolbox_editor_view.js]: console.warn or throw "unimplemented ..."

File: src/workspacefactory/wfactory_model.js [hunk 23]
```diff
@@ -378,7 +378,7 @@ WorkspaceFactoryModel.prototype.addCustomTag = function(category, tag) {
  * @param {!Element} xml The XML to be saved.
  */
 WorkspaceFactoryModel.prototype.savePreloadXml = function(xml) {
```
Comment (AnmAtAnm) [inline src/workspacefactory/wfactory_model.js:380]: I have mild concern where this method is going.  If the workspace is already represented by input param `xml`, then what is this method doing? If it is only writing the actual file, maybe the responsibility lies in the app controller.

**Review by AnmAtAnm — COMMENTED** [review]
Mostly just console.warn or throw "unimplemented ..." suggestions.  Some may overlap with the PR #77.

**Comment by svouse** [comment]
Review status: 0 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[src/model/toolbox.js, line 333 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoOXTe6oHfQsLXazRET:-KoOXTe6oHfQsLXazREU:b-hd7jp5) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L333)):*
> ```JavaScript
>    */
>   addShadowBlock(blockId) {
>     // Moved in from wfactory_model.js
> ```

Is there somewhere that you check whether or not the ID is unique?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/80)*

**Comment by chynu** [comment]
Review status: 0 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 111 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkK-r1-111:-KoOSPCFTy3jmIEll0Lm:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/controller/toolbox_controller.js#L111)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Looking at what this does in the production dev tools, let's rename this addCategorySeparator() (and update the doc).

FYI: You can also have separators between blocks within a toolbox category flyout.  Added issue #81 to track this issue (easy stretch goal).
</blockquote></details>

Done.

---

*[src/controller/toolbox_controller.js, line 125 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkK-r1-125:-KoOSU8HEX0jBkhE-9Ux:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/controller/toolbox_controller.js#L125)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Console.warn("unimplemented ToolboxController.addSeparator()")
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 105 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-105:-KoOaSt9lRDqUW0-pYrc:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L105)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 122 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-122:-KoOaTcVkbsR4KKH-iIv:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L122)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented"
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-138:-KoOTcKy9qqSSSAT4LqX:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L138)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throws "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 152 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-152:-KoOTcmEpoGtl0pS_aGc:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L152)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 173 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-173:-KoOTdLHo_F4VxOra9c_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L173)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 188 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-188:-KoOTdo0XxMIf3hDzVBG:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L188)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 203 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-203:-KoOTeDlKEVeCRHPb_FK:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L203)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-225:-KoOTeqLg3KwoyXb3slk:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L225)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 236 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-236:-KoOTfo3CRuWsctLrFOu:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L236)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 249 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-249:-KoOTgMQ5yDIXib12_AY:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L249)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 260 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-260:-KoOTgpWhfRBOjoS85Bo:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L260)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 276 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-276:-KoOThHwLVfNPrE5FJ1W:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L276)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 290 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-290:-KoOThfgythFSg0iuaUt:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L290)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 306 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-306:-KoOTi2HcIJq_vWoRzIL:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L306)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 325 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-325:-KoOTjZvMP44WiGqdXbB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L325)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 333 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoOXTe6oHfQsLXazRET:-KoOafaA9-IsnOJhcoCT:btrs30y) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L333)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Is there somewhere that you check whether or not the ID is unique?
</blockquote></details>

Done (within the function).

---

*[src/model/toolbox.js, line 349 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-349:-KoOTk65BnC7iz3-CZ2_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L349)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 364 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-364:-KoOTk_k9_n6hRUIBQOA:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L364)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 428 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkN-r1-428:-KoOTlJTIU1hiUHiqaXR:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/model/toolbox.js#L428)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/view/toolbox_editor_view.js, line 224 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkQ-r1-224:-KoO_d4AK_yTyhE5TUgW:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/view/toolbox_editor_view.js#L224)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

JSDoc
</blockquote></details>

Done.

---

*[src/view/toolbox_editor_view.js, line 231 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkQ-r1-231:-KoOU8HF3Y8Y1PRAxp2_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/view/toolbox_editor_view.js#L231)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

console.warn or throw "unimplemented ..."
</blockquote></details>

Done.

---

*[src/workspacefactory/wfactory_model.js, line 380 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/80#-KoO4pCXlecaFnPUOdkR-r1-380:-KoOUDsRh5PJ2VNPZpkG:bot1680) ([raw file](https://github.com/google/blockly-devtools/blob/29690075da7bb865ea32bcbdd55b9b3918f73fb5/src/workspacefactory/wfactory_model.js#L380)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I have mild concern where this method is going.  If the workspace is already represented by input param `xml`, then what is this method doing? If it is only writing the actual file, maybe the responsibility lies in the app controller.
</blockquote></details>

Will probably leave this function as it is for now--I think they meant for the model to keep track of what blocks were in the editor workspace so that they could re-load it when switching between Toolbox and Workspace editing tabs (in the current DevTools).

For the refactor, the `WorkspaceContents` object will store blocks as XML and the controller will retrieve the XML from the object when displaying it onto the editor workspace. (Not sure if this answers the concern.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/80)*

**Comment by AnmAtAnm** [comment]
:lgtm:

---

Review status: 0 of 9 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/80#-:-KoOdH6SdYLWpKmOP1sk:bnfp4nl)*

**Comment by svouse** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/80#-:-KoOdds4BBul5utvJMAm:bnfp4nl)*

**Review by AnmAtAnm — APPROVED** [review]

**Review by svouse — APPROVED** [review]

**Pull request merged**